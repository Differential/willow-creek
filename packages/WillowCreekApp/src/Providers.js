import React from 'react';
import Config from 'react-native-config';
import { Providers } from '@apollosproject/ui-kit';
import { AuthProvider } from '@apollosproject/ui-auth';
import { AnalyticsProvider } from '@apollosproject/ui-analytics';
import { MediaPlayerProvider } from '@apollosproject/ui-media-player';
import { NotificationsProvider } from '@apollosproject/ui-notifications';

import NavigationService from './NavigationService';
import ClientProvider from './client';
import customTheme, { customIcons } from './theme';

const AppProviders = (props) => (
  <ClientProvider {...props}>
    <NotificationsProvider
      oneSignalKey={Config.ONE_SIGNAL_KEY}
      navigate={NavigationService.navigate}
    >
      <AuthProvider
        navigateToAuth={() => NavigationService.navigate('Auth')}
        closeAuth={() => NavigationService.navigate('Onboarding')}
      >
        <MediaPlayerProvider>
          <AnalyticsProvider>
            <Providers
              themeInput={customTheme}
              iconInput={customIcons}
              {...props}
            />
          </AnalyticsProvider>
        </MediaPlayerProvider>
      </AuthProvider>
    </NotificationsProvider>
  </ClientProvider>
);

export default AppProviders;
