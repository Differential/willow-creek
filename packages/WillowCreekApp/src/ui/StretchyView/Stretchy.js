import React, { Component } from 'react';
import { Animated } from 'react-native';
import PropTypes from 'prop-types';

class Stretchy extends Component {
  static propTypes = {
    scrollY: PropTypes.shape({
      interpolate: PropTypes.func,
    }),
    x: PropTypes.number,
    y: PropTypes.number,
    width: PropTypes.number,
    height: PropTypes.number,
    layoutHeight: PropTypes.number,
    children: PropTypes.node,
    stretchOn: PropTypes.oneOf(['top', 'bottom']),
    style: PropTypes.any, // eslint-disable-line
  };

  get stretchyYPosition() {
    if (!this.props.scrollY || !this.props.height) return 0;

    if (this.props.stretchOn === 'top') {
      return this.props.scrollY.interpolate({
        inputRange: [
          this.props.y - this.props.height,
          this.props.y,
          this.props.y + this.props.height,
        ],
        outputRange: [this.props.height / 2, 0, -this.props.height],
      });
    }

    if (this.props.stretchOn === 'bottom') {
      return this.props.scrollY.interpolate({
        inputRange: [
          0,
          this.props.y + this.props.height - this.props.layoutHeight,
          this.props.y +
            this.props.height -
            this.props.layoutHeight +
            this.props.height,
        ],
        outputRange: [
          this.props.y,
          this.props.layoutHeight - this.props.height,
          this.props.layoutHeight - this.props.height - this.props.height / 2,
        ],
      });
    }

    return 0;
  }

  get stretchyScale() {
    if (!this.props.scrollY) return 1;

    if (this.props.stretchOn === 'top') {
      return this.props.scrollY.interpolate({
        inputRange: [
          this.props.y - this.props.height,
          this.props.y,
          this.props.y + this.props.height,
        ],
        outputRange: [2, 1, 1],
      });
    }

    if (this.props.stretchOn === 'bottom') {
      return this.props.scrollY.interpolate({
        inputRange: [
          0,
          this.props.y + this.props.height - this.props.layoutHeight,
          this.props.y +
            this.props.height -
            this.props.layoutHeight +
            this.props.height,
        ],
        outputRange: [1, 1, 2],
      });
    }

    return 1;
  }

  get stretchyOpacity() {
    if (!this.props.scrollY || !this.props.height) return 0;

    if (this.props.stretchOn === 'top') {
      return this.props.scrollY.interpolate({
        inputRange: [this.props.y - 0.1, this.props.y, this.props.y + 0.1],
        outputRange: [1, 0, 0],
      });
    }
    if (this.props.stretchOn === 'bottom') {
      return this.props.scrollY.interpolate({
        inputRange: [
          this.props.y + this.props.height - this.props.layoutHeight - 0.1,
          this.props.y + this.props.height - this.props.layoutHeight,
          this.props.y + this.props.height - this.props.layoutHeight + 0.1,
        ],
        outputRange: [0, 0, 1],
      });
    }

    return 0;
  }

  get stretchyStyle() {
    return [
      {
        position: 'absolute',
        top: 0,
        left: 0,
        width: this.props.width,
        height: this.props.height,
        transform: [
          { translateY: this.stretchyYPosition },
          { scale: this.stretchyScale },
        ],
        opacity: this.stretchyOpacity,
      },
      this.props.style,
    ];
  }

  render() {
    const { scrollY, style, ...otherProps } = this.props;
    return <Animated.View {...otherProps} style={this.stretchyStyle} />;
  }
}

export default Stretchy;
