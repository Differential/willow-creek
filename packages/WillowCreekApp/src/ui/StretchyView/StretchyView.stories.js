/* eslint-disable react-native/no-inline-styles */
import React from 'react';
import { storiesOf } from '@apollosproject/ui-storybook';
import { StyleSheet, ScrollView, View } from 'react-native';
import {
  ConnectedImage,
  GradientOverlayImage,
  PaddedView,
  H1,
  H2,
  H3,
} from '@apollosproject/ui-kit';

import StretchyView from '.';

const imageStyle = { width: '100%', aspectRatio: 1 };

storiesOf('StretchyView', module)
  .add('stretchy as a background', () => (
    <StretchyView>
      {({ Stretchy, ...scrollViewProps }) => (
        <ScrollView {...scrollViewProps}>
          <Stretchy background>
            <GradientOverlayImage
              overlayColor="white"
              overlayType="featured"
              source={{ uri: 'https://picsum.photos/id/233/600/600' }}
              style={imageStyle}
            />
          </Stretchy>
          <PaddedView>
            <H1 padded>Well hello there!</H1>
            <H2 padded>
              Hello there!
              {'\n'}
            </H2>
            <H3 padded>Hello there!</H3>
            <H1 padded>Try to scroll me around.</H1>
          </PaddedView>
        </ScrollView>
      )}
    </StretchyView>
  ))
  .add('stretchy as a header', () => (
    <StretchyView>
      {({ Stretchy, ...scrollViewProps }) => (
        <ScrollView {...scrollViewProps}>
          <View>
            <Stretchy>
              <ConnectedImage
                style={imageStyle}
                source={{ uri: 'https://picsum.photos/id/342/600/600' }}
              />
            </Stretchy>
            <View
              style={{
                ...StyleSheet.absoluteFillObject,
                alignItems: 'center',
                justifyContent: 'center',
              }}
            >
              <H2>Text on top of the stretchy!</H2>
            </View>
          </View>
          <PaddedView>
            <H1 padded>Well hello there!</H1>
            <H2 padded>
              Hello there!
              {'\n'}
            </H2>
            <H3 padded>Hello there!</H3>
            <H1 padded>Try to scroll me around.</H1>
          </PaddedView>
        </ScrollView>
      )}
    </StretchyView>
  ))
  .add('stretchy as a footer', () => (
    <StretchyView>
      {({ Stretchy, ...scrollViewProps }) => (
        <ScrollView {...scrollViewProps}>
          <PaddedView>
            <H1 padded>Well hello there!</H1>
            <H2 padded>
              Hello there!
              {'\n'}
            </H2>
            <H3 padded>Hello there!</H3>
            <H1 padded>Try to scroll me around.</H1>
          </PaddedView>
          <Stretchy stretchOn="bottom">
            <ConnectedImage
              style={imageStyle}
              source={{ uri: 'https://picsum.photos/id/342/600/600' }}
            />
          </Stretchy>
        </ScrollView>
      )}
    </StretchyView>
  ))
  .add('multiple stretchies', () => (
    <StretchyView>
      {({ Stretchy, ...scrollViewProps }) => (
        <ScrollView {...scrollViewProps}>
          <Stretchy stretchyKey="header" stretchOn="top">
            <ConnectedImage
              style={imageStyle}
              source={{ uri: 'https://picsum.photos/id/342/600/600' }}
            />
          </Stretchy>
          <PaddedView>
            <H1 padded>Well hello there!</H1>
            <H2 padded>
              Hello there!
              {'\n'}
            </H2>
            <H3 padded>Hello there!</H3>
            <H1 padded>Try to scroll me around.</H1>
          </PaddedView>
          <Stretchy stretchyKey="footer" stretchOn="bottom">
            <ConnectedImage
              style={imageStyle}
              source={{ uri: 'https://picsum.photos/id/342/600/600' }}
            />
          </Stretchy>
        </ScrollView>
      )}
    </StretchyView>
  ));
