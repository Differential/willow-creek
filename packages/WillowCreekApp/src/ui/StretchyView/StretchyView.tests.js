import React from 'react';
import renderer from 'react-test-renderer';

import { ScrollView, View } from 'react-native';
import {
  ConnectedImage,
  GradientOverlayImage,
  PaddedView,
  H1,
  H2,
  H3,
} from '@apollosproject/ui-kit';
import Providers from '../Providers';

import StretchyView from '.';

const imageStyle = { width: '100%', aspectRatio: 1 };

describe('the StretchyView component', () => {
  it('should render as a background', () => {
    const tree = renderer.create(
      <Providers>
        <StretchyView
          StretchyComponent={
            <GradientOverlayImage
              overlayColor="white"
              overlayType="featured"
              source={{ uri: 'https://picsum.photos/600/600' }}
              style={imageStyle}
            />
          }
        >
          {(scrollViewProps) => (
            <ScrollView {...scrollViewProps}>
              <PaddedView>
                <H1 padded>Well hello there!</H1>
                <H2 padded>
                  Hello there!
                  {'\n'}
                </H2>
                <H3 padded>Hello there!</H3>
                <H1 padded>Try to scroll me around.</H1>
              </PaddedView>
            </ScrollView>
          )}
        </StretchyView>
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
  it('should render as a header', () => {
    const tree = renderer.create(
      <Providers>
        <StretchyView
          StretchyComponent={
            <ConnectedImage
              style={imageStyle}
              source={{ uri: 'https://picsum.photos/600/600' }}
            />
          }
        >
          {({ stretchyHeight, ...scrollViewProps }) => (
            <ScrollView {...scrollViewProps}>
              <View
                style={{
                  height: stretchyHeight,
                }}
              >
                <H2>Text on top of the stretchy!</H2>
              </View>
              <PaddedView>
                <H1 padded>Well hello there!</H1>
                <H2 padded>
                  Hello there!
                  {'\n'}
                </H2>
                <H3 padded>Hello there!</H3>
                <H1 padded>Try to scroll me around.</H1>
              </PaddedView>
            </ScrollView>
          )}
        </StretchyView>
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
});
