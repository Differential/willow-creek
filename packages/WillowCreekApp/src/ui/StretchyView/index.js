import React, { PureComponent } from 'react';
import { View, StyleSheet, Animated, Dimensions } from 'react-native';
import PropTypes from 'prop-types';

import Stretchy from './Stretchy';

class StretchyView extends PureComponent {
  static propTypes = {
    children: PropTypes.func.isRequired,
  };

  state = {
    stretchies: {},
    layoutHeight: Dimensions.get('window').height,
  };

  scrollY = new Animated.Value(0);

  handleScroll = Animated.event(
    [{ nativeEvent: { contentOffset: { y: this.scrollY } } }],
    { useNativeDriver: true }
  );

  handleLayout = ({
    nativeEvent: {
      layout: { height },
    },
  }) => {
    this.setState({ layoutHeight: height });
  };

  handleStretchyLayout = ({ stretchyKey, ...props }) => {
    this.setState(({ stretchies }) => ({
      stretchies: { ...stretchies, [stretchyKey]: props },
    }));
  };

  StretchyPortal = ({
    children,
    stretchyKey = 'stretchy',
    stretchOn = 'top',
    background = false,
    style,
    ...otherProps
  }) => (
    <Animated.View
      onLayout={({ nativeEvent: { layout } }) =>
        this.handleStretchyLayout({
          stretchyKey,
          stretchOn,
          children,
          style,
          ...otherProps,
          ...layout,
        })
      }
      style={[
        { opacity: this.getStretchyPortalOpacity(stretchyKey) },
        background ? StyleSheet.absoluteFill : null,
        style,
      ]}
    >
      {children}
    </Animated.View>
  );

  getStretchyPortalOpacity = (key) => {
    if (!this.state.stretchies[key]) return 1;
    const { stretchOn, y, height } = this.state.stretchies[key];

    if (stretchOn === 'top') {
      return this.scrollY.interpolate({
        inputRange: [y - 0.1, y, y + 0.1],
        outputRange: [0, 1, 1],
      });
    }

    if (stretchOn === 'bottom') {
      return this.scrollY.interpolate({
        inputRange: [
          y + height - this.state.layoutHeight - 0.1,
          y + height - this.state.layoutHeight,
          y + height - this.state.layoutHeight + 0.1,
        ],
        outputRange: [1, 1, 0],
      });
    }

    return 1;
  };

  render() {
    const { children, ...otherProps } = this.props;

    return (
      <View
        style={StyleSheet.absoluteFill}
        {...otherProps}
        onLayout={this.handleLayout}
      >
        {Object.keys(this.state.stretchies).map((key) => (
          <Stretchy
            key={key}
            layoutHeight={this.state.layoutHeight}
            scrollY={this.scrollY}
            {...this.state.stretchies[key]}
          />
        ))}
        {children({
          scrollEventThrottle: 16,
          onScroll: this.handleScroll,
          Stretchy: this.StretchyPortal,
        })}
      </View>
    );
  }
}

export default StretchyView;
