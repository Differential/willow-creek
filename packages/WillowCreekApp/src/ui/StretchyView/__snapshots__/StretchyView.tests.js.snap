// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the StretchyView component should render as a background 1`] = `
<View
  StretchyComponent={
    <pure(Component)
      maintainAspectRatio={true}
      overlayColor="white"
      overlayType="featured"
      source={
        Object {
          "uri": "https://picsum.photos/600/600",
        }
      }
      style={
        Object {
          "aspectRatio": 1,
          "width": "100%",
        }
      }
    />
  }
  onLayout={[Function]}
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <RCTScrollView
    Stretchy={[Function]}
    onScroll={[Function]}
    scrollEventThrottle={16}
  >
    <View>
      <View
        style={
          Object {
            "paddingHorizontal": 16,
            "paddingVertical": 16,
          }
        }
      >
        <Text
          padded={true}
          style={
            Object {
              "color": "#303030",
              "fontFamily": "InterUI-Black",
              "fontSize": 43,
              "lineHeight": 49.45,
              "paddingVertical": 48.38,
            }
          }
        >
          Well hello there!
        </Text>
        <Text
          padded={true}
          style={
            Object {
              "color": "#303030",
              "fontFamily": "InterUI-Black",
              "fontSize": 36,
              "lineHeight": 41.4,
              "paddingBottom": 20.02,
              "paddingTop": 24,
            }
          }
        >
          Hello there!
          

        </Text>
        <Text
          padded={true}
          style={
            Object {
              "color": "#303030",
              "fontFamily": "InterUI-Black",
              "fontSize": 24,
              "lineHeight": 27.6,
              "paddingBottom": 18,
              "paddingTop": 27,
            }
          }
        >
          Hello there!
        </Text>
        <Text
          padded={true}
          style={
            Object {
              "color": "#303030",
              "fontFamily": "InterUI-Black",
              "fontSize": 43,
              "lineHeight": 49.45,
              "paddingVertical": 48.38,
            }
          }
        >
          Try to scroll me around.
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`the StretchyView component should render as a header 1`] = `
<View
  StretchyComponent={
    <mapProps(getContext(mapProps(withPropsOnChange(mapProps(ConnectedImage)))))
      source={
        Object {
          "uri": "https://picsum.photos/600/600",
        }
      }
      style={
        Object {
          "aspectRatio": 1,
          "width": "100%",
        }
      }
    />
  }
  onLayout={[Function]}
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <RCTScrollView
    Stretchy={[Function]}
    onScroll={[Function]}
    scrollEventThrottle={16}
  >
    <View>
      <View
        style={
          Object {
            "height": undefined,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#303030",
              "fontFamily": "InterUI-Black",
              "fontSize": 36,
              "lineHeight": 41.4,
            }
          }
        >
          Text on top of the stretchy!
        </Text>
      </View>
      <View
        style={
          Object {
            "paddingHorizontal": 16,
            "paddingVertical": 16,
          }
        }
      >
        <Text
          padded={true}
          style={
            Object {
              "color": "#303030",
              "fontFamily": "InterUI-Black",
              "fontSize": 43,
              "lineHeight": 49.45,
              "paddingVertical": 48.38,
            }
          }
        >
          Well hello there!
        </Text>
        <Text
          padded={true}
          style={
            Object {
              "color": "#303030",
              "fontFamily": "InterUI-Black",
              "fontSize": 36,
              "lineHeight": 41.4,
              "paddingBottom": 20.02,
              "paddingTop": 24,
            }
          }
        >
          Hello there!
          

        </Text>
        <Text
          padded={true}
          style={
            Object {
              "color": "#303030",
              "fontFamily": "InterUI-Black",
              "fontSize": 24,
              "lineHeight": 27.6,
              "paddingBottom": 18,
              "paddingTop": 27,
            }
          }
        >
          Hello there!
        </Text>
        <Text
          padded={true}
          style={
            Object {
              "color": "#303030",
              "fontFamily": "InterUI-Black",
              "fontSize": 43,
              "lineHeight": 49.45,
              "paddingVertical": 48.38,
            }
          }
        >
          Try to scroll me around.
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
